lib/autokey/scripting/engine.py

```diff
@@ -122,12 +168,90 @@ def create_phrase(self, folder, name: str, contents: str,
                                     If set to True, the new Phrase will be shown in the tray icon context menu.
         @param always_prompt: A boolean defaulting to False. If set to True,
                               the Phrase expansion has to be manually confirmed, each time it is triggered.
+        @param temporary: Hotkeys created with temporary=True are
+                                    not persisted as .jsons, and are replaced if the description is not
+                                    unique within the folder.
+                                     Used for single-source rc-style scripts.
         @raise ValueError: If a given abbreviation or hotkey is already in use or parameters are otherwise invalid
         @return The created Phrase object. This object is NOT considered part of the public API and exposes the raw
           internals of AutoKey. Ignore it, if you don’t need it or don’t know what to do with it.
           It can be used for _really_ advanced use cases, where further customizations are desired. Use at your own
           risk. No guarantees are made about the object’s structure. Read the AutoKey source code for details.
         """
+        # Start with input type-checking.
+        if not isinstance(folder, model.Folder):
+            raise ValueError("Expected a folder, not {}".format(
+                type(folder))
+            )
+        if not isinstance(name, str):
+            raise ValueError("Expected name to be str, not {}".format(
+                type(name))
+            )
+        if not isinstance(contents, str):
+            raise ValueError("Expected contents to be str, not {}".format(
+                type(contents))
+            )
+        if abbreviations is not None:
+            fail=False
+            if not isinstance(abbreviations, str):
+                fail=True
+                if isinstance(abbreviations, list):
+                    fail=False
+                    for item in abbreviations:
+                        if not isinstance(item, str):
+                            fail=True
+            if fail:
+                raise ValueError("Expected abbreviations to be str or list of str, not {}".format(
+                    type(abbreviations))
+                    )
+        if hotkey is not None:
+            fail=False
+            if isinstance(hotkey, tuple):
+                if len(hotkey) == 2:
+                    # First check modifiers is list of str or keys
+                    if isinstance(hotkey[0], list):
+                        for item in hotkey[0]:
+                            if not isinstance(item, str) and not \
+                            isinstance(item, model.Key):
+                                fail=True
+                    elif not isinstance(hotkey[0], str) and not \
```

Comment:
Just about the readability:
Personally, these instance checks are incredibly hard to parse for me. I don’t know if it’s just me, but it took me over two minutes to think about these lines and check if this is actually correct. I solved it by applying De-Morgan rule which made it immediately obvious.
I think `not(isinstance(item, str) or isinstance(item, model.Key))` is way easier to read, as "is not one of two possible types".

So I think I’ll let my IDE apply De-Morgan rule here.

Reply:
Yeah, I struggled with this one. Also would be happy turning these into functions, eg for validating parent folders for create_folder.
There's a test made, but the hotkey one isn't totally comprehensive I don't think

Reply:
Yeah, I struggled with this one. Also would be happy turning these into functions, eg for validating parent folders for create_folder.
There's a test made, but the hotkey one isn't totally comprehensive I don't think